### PR TITLE
Adding OpenStack Swift support to blocks storage

### LINKED ISSUE #2861: Add OpenStack Swift support to blocks storage
Thanos supports OpenStack Swift so we could easily support it too for the blocks storage, but currently lack some config mapping and bucket client initialisation code. We should do it.

### PR DESCRIPTION
**Pull request by deniszh**
**What this PR does**:
Adding OpenStack Swift support to blocks storage

**Which issue(s) this PR fixes**:
Fixes #2861 

**Checklist**
- [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

Notes: 
 * I included compose setup with docker swift image to test it, but it's unfortunately not automatically working. `cortex-rules` container is creating by Cortex itself, but `cortex-tsdb` is not. So, script "create_container.sh" needs to be run once after starting compose. Also compose setup working only in docker host network mode (calling swift on localhost), otherwise swift container can't properly resolve URL

### CODE REVIEW

**Comment by deniszh** [comment]
Not sure how to fix
```
pkg/storage/backend/swift/config.go:9:2: struct field `AuthUrl` should be `AuthURL` (golint)
	AuthUrl           string `yaml:"auth_url"`
	^
pkg/storage/backend/swift/config.go:13:2: struct field `UserId` should be `UserID` (golint)
	UserId            string `yaml:"user_id"`
	^
pkg/storage/backend/swift/config.go:15:2: struct field `DomainId` should be `DomainID` (golint)
	DomainId          string `yaml:"domain_id"`
	^
```
:(

**Comment by deniszh** [comment]
Also not sure why website build failed with some scary error message.

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:45]:
```suggestion
* [FEATURE] Add OpenStack Swift support to blocks storage. #3303
```

File: pkg/storage/backend/swift/config.go
Comment (pracucci) [inline pkg/storage/backend/swift/config.go:31]: Few things, please:
1. Capitalise `OpenStack` in CLI flags description. Also I would refer to it as `OpenStack Swift` and not just `OpenStack`
2. Make sure each CLI flag description ends with a dot `.`
3. `id` > `ID`

File: pkg/storage/backend/swift/config.go
Comment (pracucci) [inline pkg/storage/backend/swift/config.go:45]: A couple of things:
1. Default is `""` but the description says `(default "cortex")`
2. Do not specify what's the default in the CLI flag description, because when we build the help/doc we'll use the default actually specified in the code (which right now is `""`)

File: pkg/storage/backend/swift/config.go
Comment (pracucci) [inline pkg/storage/backend/swift/config.go:44]: > default is use first region

Could you clarify what this mean?

File: pkg/storage/backend/swift/config.go
Comment (pracucci) [inline pkg/storage/backend/swift/config.go:33]:
```suggestion
	f.StringVar(&cfg.Username, prefix+"swift.username", "", "OpenStack Swift username.")
```

File: pkg/storage/backend/swift/config.go
Comment (pracucci) [inline pkg/storage/backend/swift/config.go:36]: Could you also clarify the difference between "username", "user domain" and "user ID"? Are they mutually exclusive or not? When should be set?

```suggestion
	f.StringVar(&cfg.UserId, prefix+"swift.user-id", "", "OpenStack Switch user ID.")
```

**Review by pracucci — COMMENTED** [review]
Thanks for working on this! Much appreciated. I left few comments to move forward: could you take a look, please? Thanks! 👏

File: pkg/storage/backend/swift/config.go
Comment (deniszh) [reply pkg/storage/backend/swift/config.go:44]: I mostly copied description from ncw/swift docs - https://github.com/ncw/swift/blob/master/swift.go#L96
As far as I understand, region can be region name, or comma separated string - then it will use first region in the string. It's directly going to library, so, I think we can just remove this description to decrease confusion.

File: pkg/storage/backend/swift/config.go
Comment (deniszh) [reply pkg/storage/backend/swift/config.go:36]: That's kinda complicated. Not sure should we put that in cli help. It's about different Swift Auth API versions. See e.g. https://godoc.org/github.com/ncw/swift 
For v1 (which is obsolete) - only URL, username and password are needed.
For V2 - URL, username, password are mandatory, then tenant name OR tenant ID is mandatory, and region is mandatory.
V3 is even more complex, all these user domain/id, project domain/id are related to it.
Usually, Swift users knows exactly which auth version is supported and which parameters needs to be provided.

File: pkg/storage/backend/swift/config.go
Comment (deniszh) [reply pkg/storage/backend/swift/config.go:31]: Done

File: pkg/storage/backend/swift/config.go
Comment (deniszh) [reply pkg/storage/backend/swift/config.go]: Let me know what's better to do here, please

**Comment by deniszh** [comment]
Now docker compose is working out of the box.

**Comment by pracucci** [comment]
> Not sure how to fix
> 
> ```
> pkg/storage/backend/swift/config.go:9:2: struct field `AuthUrl` should be `AuthURL` (golint)
> 	AuthUrl           string `yaml:"auth_url"`
> 	^
> pkg/storage/backend/swift/config.go:13:2: struct field `UserId` should be `UserID` (golint)
> 	UserId            string `yaml:"user_id"`
> 	^
> pkg/storage/backend/swift/config.go:15:2: struct field `DomainId` should be `DomainID` (golint)
> 	DomainId          string `yaml:"domain_id"`
> 	^
> ```
> 
> becuse they exported like this from https://github.com/thanos-io/thanos/blob/master/pkg/objstore/swift/swift.go#L30-L45
> :(

This config struct is defined in cortex. Just rename these fields using the case suggested by the linter.

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:52]:
```suggestion
* [FEATURE] Add OpenStack Swift support to blocks storage. #3303
```

File: pkg/storage/backend/swift/config.go
Comment (pracucci) [inline pkg/storage/backend/swift/config.go:36]:
```suggestion
	f.StringVar(&cfg.UserId, prefix+"swift.user-id", "", "OpenStack Swift user ID.")
```

File: pkg/storage/backend/swift/config.go
Comment (pracucci) [inline pkg/storage/backend/swift/config.go:37]:
```suggestion
	f.StringVar(&cfg.Password, prefix+"swift.password", "", "OpenStack Swift API key.")
```

File: pkg/storage/backend/swift/config.go
Comment (pracucci) [inline pkg/storage/backend/swift/config.go:40]:
```suggestion
	f.StringVar(&cfg.ProjectID, prefix+"swift.project-id", "", "OpenStack Swift project ID (v2,v3 auth only).")
```

**Review by pracucci — APPROVED** [review]
Very good job! LGTM. I just left few last suggestions. Please also remember to `make doc` once CLI flag description changes are done.

**Comment by deniszh** [comment]
Integration test failed:
```
18:04:50 purger: level=error ts=2020-10-15T18:04:50.328051302Z caller=purger.go:240 user_id=user-1 request_id=f4c76923 msg="error removing delete plan" plan_no=2 err="remove /shared/user-1:f4c76923: no such file or directory"
    chunks_delete_series_test.go:182: 
        	Error Trace:	chunks_delete_series_test.go:182
        	Error:      	Received unexpected error:
        	            	unable to find metrics [cortex_purger_delete_requests_processed_total] with expected values. Last error: <nil>. Last values: [2]
        	Test:       	TestDeleteSeriesAllIndexBackends
```
Looks unrelated, but I may be wrong.

**Comment by pracucci** [comment]
Thanks @deniszh! LGTM, good job! I pinged another maintainer for the 2nd review and then we'll merge.

**Comment by deniszh** [comment]
@pstibrany: Could you please review it when you have time? Thanks a lot!

**Comment by deniszh** [comment]
FYI - thanos has PR for refactor OpenSwift support - https://github.com/thanos-io/thanos/pull/2732
I'm following it and will change this PR or fix it if needed.

**Comment by pracucci** [comment]
> FYI - thanos has PR for refactor OpenSwift support - [thanos-io/thanos#2732](https://github.com/thanos-io/thanos/pull/2732)
> I'm following it and will change this PR or fix it if needed.

I would keep this PR as is. You can then work on a follow-up PR once that Thanos PR will be merged.

**Review by pracucci — APPROVED** [review]

**Review by pstibrany — APPROVED** [review]
LGTM, thank you, and apologies for slow review.

**Pull request merged**
